### PR TITLE
Add check for duplicate variables in binding patterns

### LINKED ISSUE #17930: [Spec Deviation] Variables in a binding-pattern must be distinct
**Description:**
According to the 2019r3 spec, 
```
All the variables in a binding-pattern must be distinct e.g. [x, x] is not allowed.
```

The implementation does not enforce this.

**Steps to reproduce:**
The following works:
```ballerina
import ballerina/io;

public function main() {
    [int, int, int...] x = [1, 2, 3, 4];

    int a;
    int[] b; 

    [a, a, ...b] = x; // note how `a` is repeated
    io:println(a);
    io:println(b);
}
```

**Affected Versions:**
1.0.0-beta-SNAPSHOT - 15th August build

### PR DESCRIPTION
**Pull request by grainier**
## Purpose
$subject

Fixes #17930 

## Remarks
https://github.com/ballerina-platform/ballerina-lang/issues/17930

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 1]
```diff
@@ -1265,24 +1266,61 @@ public void visit(BLangAssignment assignNode) {
...
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
     @Override
     public void visit(BLangTupleDestructure stmt) {
+        List<BLangExpression> varRefs = stmt.varRef.expressions.stream()
+                .filter(e -> e instanceof BLangVariableReference)
```
Comment (Anoukh) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Is there any way we can remove instanceof check?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 2]
```diff
@@ -1265,24 +1266,61 @@ public void visit(BLangAssignment assignNode) {
...
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
+    private void checkDuplicateVarRefs(List<BLangExpression> varRefs) {
+        Set<BSymbol> symbols = new HashSet<>();
+        for (BLangExpression varRef : varRefs) {
+            if (varRef != null && (varRef.getKind() == NodeKind.SIMPLE_VARIABLE_REF
```
Comment (Anoukh) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Shall we check if we can invert the if condition?

Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: fixed in 102c92a

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 1 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: It's redundant and removed in 102c92a

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 3]
```diff
@@ -1265,24 +1266,62 @@ public void visit(BLangAssignment assignNode) {
...
+        this.checkDuplicateVarRefs(varRefs);
         this.checkStatementExecutionValidity(stmt);
         analyzeExpr(stmt.varRef);
         analyzeExpr(stmt.expr);
     }
 
+    private void checkDuplicateVarRefs(List<BLangExpression> varRefs) {
+        Set<BSymbol> symbols = new HashSet<>();
+        for (BLangExpression varRef : varRefs) {
+            if (varRef == null || (varRef.getKind() != NodeKind.SIMPLE_VARIABLE_REF
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: This logic breaks, If I have nested binding patterns. i.e `{ message : x , data : [ x, y] } = someRecord();`

Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: fixed in f6dd911

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 4]
```diff
@@ -1265,24 +1266,89 @@ public void visit(BLangAssignment assignNode) {
...
+    private void checkDuplicateVarRefs(List<BLangExpression> varRefs) {
+        checkDuplicateVarRefs(varRefs, new HashSet<>());
+    }
+
+    private void checkDuplicateVarRefs(List<BLangExpression> varRefs, Set<BSymbol> symbols) {
+        for (BLangExpression varRef : varRefs) {
+            if (varRef == null || (varRef.getKind() != NodeKind.SIMPLE_VARIABLE_REF
+                    && varRef.getKind() != NodeKind.INDEX_BASED_ACCESS_EXPR
+                    && varRef.getKind() != NodeKind.FIELD_BASED_ACCESS_EXPR
+                    && varRef.getKind() != NodeKind.XML_ATTRIBUTE_ACCESS_EXPR
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: NodeKind.INDEX_BASED_ACCESS_EXPR and NodeKind.FIELD_BASED_ACCESS_EXPR and XML_ATTRIBUTE_ACCESS_EX are not allowed according to the spec.  Let's fix them in a separate PR.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 5]
```diff
@@ -1265,24 +1266,89 @@ public void visit(BLangAssignment assignNode) {
     }
 
     public void visit(BLangRecordDestructure stmt) {
+        List<BLangExpression> varRefs = stmt.varRef.recordRefFields.stream()
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: The same logic is repeated in later. I think we can simplify this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 4 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Created https://github.com/ballerina-platform/ballerina-lang/issues/18249

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 5 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: fixed in 1bf5446

**Pull request merged**